test/libevmasm/PlainAssemblyParser.cpp

```diff
@@ -39,16 +37,49 @@ using namespace solidity::langutil;
 
 Json PlainAssemblyParser::parse(std::string _sourceName, std::string const& _source)
 {
+	m_sourceStream = std::istringstream(_source);
 	m_sourceName = std::move(_sourceName);
-	Json codeJSON = Json::array();
-	std::istringstream sourceStream(_source);
-	while (getline(sourceStream, m_line))
+	m_lineNumber = 0;
+
+	advanceLine();
+	return parseAssembly(0);
+}
+
+Json PlainAssemblyParser::parseAssembly(size_t _nestingLevel)
+{
+	Json assemblyJSON = {{".code", Json::array()}};
+	Json& codeJSON = assemblyJSON[".code"];
+
+	while (m_line.has_value())
 	{
-		advanceLine(m_line);
 		if (m_lineTokens.empty())
+		{
+			advanceLine();
 			continue;
+		}
+
+		size_t newLevel = parseNestingLevel();
+		if (newLevel > _nestingLevel)
+			BOOST_THROW_EXCEPTION(std::runtime_error(formatError("Indentation does not match the current subassembly nesting level.")));
 
-		if (c_instructions.contains(currentToken().value))
+		if (newLevel < _nestingLevel)
+			return assemblyJSON;
+
+		if (currentToken().value == ".sub")
+		{
+			advanceLine();
+
+			if (!assemblyJSON.contains(".data"))
+				assemblyJSON[".data"] = Json::object();
```

Comment:
```suggestion
```

Removing this changes behavior slightly: `j[".data"];` leads to `{".data": null}`; `j[".data"] = Json::object();` leads to `{".data": {}}`. Both reports size to be zero and once `nextDataIndex` is populated as child of `.data`, the null value gets replaced by something non-null.

Reply:
Good point. It will still work without this bit. Removed.

TBH I wish `[]` in C++ would generally just throw on non-existent items instead of adding them.

Reply:
Yeah - there is a bit of a zoo of stuff that can happen right? :)

- vector/array etc lead to UB
- map and friends insert default element
- bitset performs range check and throws actually

One could argue that each of these are in their own category of container so I guess it's fine but I too would prefer if it behaved uniformly. Then one didn't have to keep this in mind.